app/src/main/java/org/lineageos/tv/launcher/adapter/MainVerticalAdapter.kt

```diff
@@ -5,89 +5,58 @@
-    }
+import org.lineageos.tv.launcher.model.MainRowItem
+import org.lineageos.tv.launcher.view.MainRowItemView
+
+class MainVerticalAdapter :
+    ListAdapter<Pair<Long, MainRowItem>, MainVerticalAdapter.ViewHolder>(diffCallback) {
+    inner class ViewHolder(
+        private val mainRowItemView: MainRowItemView,
+    ) : RecyclerView.ViewHolder(mainRowItemView) {
+        fun bind(item: Pair<Long, MainRowItem>, position: Int) {
+            mainRowItemView.setData(item.second)
+            mainRowItemView.updateLayoutParams {
+                width = ViewGroup.LayoutParams.MATCH_PARENT
+                height = mainRowItemView.resources.getDimension(
+                    if (item.second.adapter is TvAdapter<*, *>) {
+                        R.dimen.main_app_row_height
+                    } else {
+                        R.dimen.main_row_height
+                    }
+                ).toInt()
+            }
 
-    fun removeItem(item: Long) {
-        for ((i, row) in rowList.withIndex()) {
-            if (row.first == item) {
-                rowList.remove(row)
-                notifyItemRemoved(i)
-                return
+            if (position == 0) {
+                mainRowItemView.requestFocus()
             }
         }
     }
 
-    fun addItem(item: Pair<Long, org.lineageos.tv.launcher.model.MainRowItem>) {
-        var temp = rowList.toMutableList()
-        temp.add(item)
-        temp =
-            temp.orderSuggestions(Suggestions.getChannelOrder(context)) { it.first } as MutableList
-        var index = temp.indexOf(item)
-        if (index == -1) {
-            index = (rowList.size - 1)
-        }
-        rowList.add(index, item)
-        notifyItemInserted(index)
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(getItem(position), position)
     }
 
-    fun isChannelShowing(channelId: Long?): Boolean {
-        channelId ?: return false
-        val res = rowList.find { it.first == channelId }
-        res ?: return false
-        return true
-    }
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ViewHolder(
+        MainRowItemView(parent.context)
+    )
 
-    fun findChannelIndex(channelId: Long?): Int {
-        channelId ?: return -1
-        return rowList.indexOfFirst { it.first == channelId }
-    }
+    companion object {
+        private val diffCallback = object : DiffUtil.ItemCallback<Pair<Long, MainRowItem>>() {
+            override fun areItemsTheSame(
+                oldItem: Pair<Long, MainRowItem>,
+                newItem: Pair<Long, MainRowItem>
+            ) = oldItem.first == newItem.first
 
-    fun itemMoved(from: Int, to: Int) {
-        Collections.swap(rowList, from, to)
-        notifyItemMoved(from, to)
+            override fun areContentsTheSame(
+                oldItem: Pair<Long, MainRowItem>,
+                newItem: Pair<Long, MainRowItem>
+            ) = oldItem.second == newItem.second
```

Comment:
this still doesn't have .equal, maybe you can check label?

Reply:
Done: Added label equals plus checking if the adapter is the same object, we handle `submitData` in the flow so makes sense to only rebind the adapter if it changed